Cython/Compiler/Nodes.py

```diff
@@ -4521,6 +4532,32 @@ def generate_argument_type_tests(self, code):
                                           arg.type.is_buffer or
                                           arg.type.is_memoryviewslice):
                 self.generate_arg_none_check(arg, code)
+        if self.target.entry.is_special:
+            for n in reversed(range(len(self.args), self.signature.max_num_fixed_args())):
+                # for special functions with optional args (e.g. power which can
+                # take 2 or 3 args), unused args are None since this is what the
+                # compilers sets
+                if self.target.entry.name == "__ipow__":
+                    # Bug in Python < 3.8 - __ipow__ is used as a binary function
+                    # and attempts to access the third argument will always fail
+                    code.putln("#if PY_VERSION_HEX >= 0x03080000")
+                code.putln("if (unlikely(unused_arg_%s != Py_None)) {" % n)
+                code.putln(
+                    'PyErr_SetString(PyExc_TypeError, '
+                    '"%s() takes %s arguments but %s were given");' % (
+                        self.target.entry.qualified_name, self.signature.max_num_fixed_args(), n))
+                code.putln("%s;" % code.error_goto(self.pos))
+                code.putln("}")
+                if self.target.entry.name == "__ipow__":
+                    code.putln("#endif /*PY_VERSION_HEX >= 0x03080000*/")
+            if self.target.entry.name == "__ipow__" and len(self.args) != 2:
+                # It's basically impossible to safely support it:
+                # Class().__ipow__(1) is guarranteed to crash.
+                # therefore, just emit a C error as the safest thing to do.
+                code.putln("#if PY_VERSION_HEX < 0x03080000")
+                code.putln('#error "%s: 3-argument %s cannot be used in Python<3.8"' % (
+                            format_position(self.pos), self.target.entry.qualified_name))
+                code.putln('#endif')
```

Comment:
Thinking about it more, not implemented would be better. I'll come back to this later... Marking as draft just so it doesn't get merged before then

Reply:
I had another look at it:
* not assigning the type slot is really hard - you just don't have the right information available when the code is generated.
* Runtime warning + NotImplemented doesn't fall back to `__pow__`.
* Therefore, I think raising an error is as good a solution as anything. Especially since this is really rare - the only way to call a 3-arg `__ipow__` is from a C API function and it isn't possible directly from Python. Therefore I don't think it's worth wasting too much more time on.

Reply:
A runtime error is ok for me.